Added find_libblis function to rocBLAS/clients CMakeLists.txt

umarinkovic opened the pull request:
## Motivation

I was building rocBLAS linked with BLIS and saw that part of the logic for searching for AOCL was hardcoded with respect to different versions. Specifically, I have the newest, 5.1.0 installed and the current logic doesn't account for that. 

## Technical Details

I wrote a function to replace the hardcoded logic by using file glob and regex patterns.

## Test Plan

Tested locally with my 5.1.0 version.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

TorreZuk commented:
@umarinkovic can you remove the formatting commit so we can more easily see the changes.   The aocl blist 5.1 release has a NaN propagation violation for gemv you can look at SWLCSG-3742, so we are still encouraging 4.2 gcc for rocblas/hipblas.  Otherwise if this works may be fine, if you can remove format that will re-run windows CI

umarinkovic commented:
> can you remove the formatting commit so we can more easily see the changes. The aocl blist 5.1 release has a NaN propagation violation for gemv you can look at SWLCSG-3742, so we are still encouraging 4.2 gcc for rocblas/hipblas. Otherwise if this works may be fine, if you can remove format that will re-run windows CI

Reverted the formatting commit.

umarinkovic commented:
@TorreZuk 

Hi! I’ve reverted the formatting commit. Could you please take another look when you have a moment?

TorreZuk commented on projects/rocblas/clients/CMakeLists.txt:
```diff
@@ -134,44 +154,40 @@ endif( )
+      set(BLIS_INCLUDE_DIR)
+      find_libblis(gcc)
+      if(NOT BLAS_LIBRARY)
+        find_libblis(aocc)
```
Right before this should be the line
set(WARN_NOT_ILP64_PREFERRED TRUE) # found via aocc
Or right after this:
   if(NOT BLAS_LIBRARY)
Not as in else below  (I will comment below at that line).
We prefer gcc build of AOCL.   We could also accept aocc as preferred we just used gcc to match other dependency toolchain rather than add another if static linking.

umarinkovic replied:
Hmm, yeah, the way it's written, if it's found via `gcc`, it's still going to be set since `if(NOT BLAS_LIBRARY)` will evaluate to false.

However setting the warning inside this block:
```cmake
if(NOT BLAS_LIBRARY)
        find_libblis(aocc)
```

makes it so that the warning is set pre-emptively and might lead to false positives (in case it isn't found in the standard aocc install location either, the following if-elif-else branches sometimes set the warning and sometimes not.

So I've added a reset of the warning in case it isn't found via standard aocc install location either, I think the logic is sound now.

TorreZuk commented on projects/rocblas/clients/CMakeLists.txt:
```diff
@@ -134,44 +154,40 @@ endif( )
         endif()
+      else()
+        set(WARN_NOT_ILP64_PREFERRED TRUE) # found via aocc
       endif()
```
This else would be WARN_NOT_ILP64_PREFERRED FALSE or just remove the else(  ) as it matches up if NOT BLAS_LIBRARY .   We are warning when not using preferred reference.

TorreZuk reviewed (commented):
Think the warning needs adjustment otherwise basics should be fine as will recommend a version in the docs.  Any reference library may have bugs but AOCL 4.2 gcc is issue free :-)   Next release of AOCL need to test again with all psdb/osdb and even stress tests.

TorreZuk reviewed (approved):
Changes look fine now, the tensile failure on windows you may need to rebase this branch as the version error suggests an old tensile

umarinkovic commented:
> Changes look fine now, the tensile failure on windows you may need to rebase this branch as the version error suggests an old tensile

Rebased on top of develop, is it okay now? I've also cleaned up the commit history a bit removing redundant commit-revert_commit-type commits

TorreZuk commented:
@umarinkovic let us let the CI testing to finish and I can ask for help merging if failures are unrelated.  We need to re-test with next release 5.2 which may come out soon.
